ui/ducks/send/send.js

```diff
@@ -0,0 +1,1238 @@
+      state.draftTransaction.txParams.from = action.payload.from;
+      slice.caseReducers.updateDraftTransaction(state);
+    },
+    /**
+     * sets the provided gasLimit in state and then recomputes the gasTotal,
+     * setting that value in state as well. Once gasTotal is computed
+     * recomputes the maximum amount if the current amount mode is 'MAX' and
+     * sending the native token. ERC20 assets max amount is unaffected by
+     * gasTotal so does not need to be recomputed. Finally, validates the gas
+     * field and send state, then updates the draft transaction.
+     */
+    updateGasLimit: (state, action) => {
+      state.gas.gasLimit = addHexPrefix(action.payload);
+      state.gas.gasTotal = addHexPrefix(
+        calcGasTotal(state.gas.gasLimit, state.gas.gasPrice),
+      );
+      if (
+        state.amount.mode === AMOUNT_MODES.MAX &&
+        state.asset.type === ASSET_TYPES.NATIVE
+      ) {
+        slice.caseReducers.updateAmountToMax(state);
+      }
+      slice.caseReducers.validateGasField(state);
+      // validate send state
+      slice.caseReducers.validateSendState(state);
+      slice.caseReducers.updateDraftTransaction(state);
+    },
+    /**
+     * sets the provided gasPrice in state and then recomputes the gasTotal,
+     * setting that value in state as well. Once gasTotal is computed
+     * recomputes the maximum amount if the current amount mode is 'MAX' and
+     * sending the native token. ERC20 assets max amount is unaffected by
+     * gasTotal so does not need to be recomputed. Finally, validates the gas
+     * field and send state, then updates the draft transaction.
+     */
+    updateGasPrice: (state, action) => {
+      state.gas.gasPrice = addHexPrefix(action.payload);
+      state.gas.gasTotal = addHexPrefix(
+        calcGasTotal(state.gas.gasLimit, state.gas.gasPrice),
+      );
+      if (
+        state.amount.mode === AMOUNT_MODES.MAX &&
+        state.asset.type === ASSET_TYPES.NATIVE
+      ) {
+        slice.caseReducers.updateAmountToMax(state);
+      }
+      slice.caseReducers.validateGasField(state);
+      // validate send state
+      slice.caseReducers.validateSendState(state);
+      slice.caseReducers.updateDraftTransaction(state);
+    },
+    /**
+     * sets the amount mode to the provided value as long as it is one of the
+     * supported modes (MAX|INPUT)
+     */
+    updateAmountMode: (state, action) => {
+      if (Object.values(AMOUNT_MODES).includes(action.payload)) {
+        state.amount.mode = action.payload;
+      }
+    },
+    updateAsset: (state, action) => {
+      state.asset.type = action.payload.type;
+      state.asset.balance = action.payload.balance;
+      if (state.asset.type === ASSET_TYPES.TOKEN) {
+        state.asset.details = action.payload.details;
+      } else {
+        // clear the details object when sending native currency
+        state.asset.details = null;
+        if (state.recipient.error === CONTRACT_ADDRESS_ERROR) {
```

Comment:
this could be called with same asset and clear this valid error

Reply:
yeah nah, this branch won't be called when updating an asset, its only when updating a native currency and in that case the errors are always invalid errors that are safe to clear.